File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-3.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int res = 0, n = heights.size();

        for(int i=0;i<n;i++) {
            int prev = i, next = i+1;
            while(prev>=0 && heights[prev] >= heights[i]) prev--;
            while(next<n && heights[next] >= heights[i]) next++;

            res = max(res, heights[i] * (next-prev-1));
        }
        return res;
    }
};

File: Data Structures & Algorithms/minimum-window-with-characters/submission-3.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        target, window = {}, {}
        for ch in t:
            target[ch] = 1 + target.get(ch, 0)
        
        need, have = len(target), 0
        res = (-1, float('inf'))
        prev = 0

        for i in range(len(s)):
            window[s[i]] = 1 + window.get(s[i], 0)
            if s[i] in target and window[s[i]] == target[s[i]]:
                have += 1
            
            while have == need:
                if res[1] > i - prev + 1:
                    res = (prev, i-prev+1)
                window[s[prev]] -= 1
                if s[prev] in target and window[s[prev]] < target[s[prev]]:
                    have -= 1
                prev += 1

        return s[res[0]:res[0]+res[1]] if res[0] != -1 else ""